hack/generator/pkg/codegen/pipeline_determine_resource_ownership.go

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+const resourcesPropertyName = astmodel.PropertyName("Resources")
+
+func determineResourceOwnership() PipelineStage {
+	return PipelineStage{
+		id:          "determineResourceOwnership",
+		description: "Determine ARM resource relationships",
+		Action:      determineOwnership,
+	}
+}
+
+func determineOwnership(ctx context.Context, definitions astmodel.Types) (astmodel.Types, error) {
+
+	updatedDefs := make(astmodel.Types)
+
+	for _, def := range definitions {
+		if resourceType, ok := def.Type().(*astmodel.ResourceType); ok {
+			specDef, err := getResourceSpecDefinition(definitions, resourceType)
+			if err != nil {
+				return nil, errors.Wrapf(err, "couldn't get spec definition for resource %s", def.Name())
+			}
+
+			specType, err := resourceSpecTypeAsObject(specDef)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Couldn't extract resource %s spec type as object", def.Name())
+			}
+
+			childResourcePropertyTypeDef, err := extractChildResourcePropertyTypeDef(
+				definitions,
+				def.Name(),
+				specDef.Name(),
+				specType)
+			if err != nil {
+				return nil, err
+			}
+			if childResourcePropertyTypeDef == nil {
+				continue // This just means skip
+			}
+
+			childResourceTypeNames, err := extractChildResourceTypeNames(*childResourcePropertyTypeDef)
+			if err != nil {
+				return nil, err
+			}
+
+			err = updateChildResourceDefinitionsWithOwner(definitions, childResourceTypeNames, def.Name(), updatedDefs)
+			if err != nil {
+				return nil, err
+			}
+
+			// Remove the resources property from the owning resource spec
+			specDef = specDef.WithType(specType.WithoutProperty(resourcesPropertyName))
+
+			updatedDefs[specDef.Name()] = specDef
+		}
+	}
+
+	setResourceGroupOwnerForResourcesWithNoOwner(definitions, updatedDefs)
+	return combineNewAndExistingDefs(definitions, updatedDefs), nil
```

Comment:
I'm not opposed, but `combineNewAndExistingDefs` actually does a few things:
1. Construct new `Types` map.
2. `AddAll` `updatedDefs` to it.
3. Add remaining `definitions` which aren't already there.

If we translated this as written using the helper functions you've added the only win we get (I think) is that the manual `AddAll` is replaced with the function cal.

We could theoretically rewrite this as:
1. Construct new `Types` map
2. `AddAll` `updatedDefs` to it.
3. Use `Where` to filter `definitions` down to only types which aren't already in the new map.
4. `AddAll` those defs.

This adds an extra loop to the process though.

Alternatively (and maybe best?) we could promote this pattern to a method on `Types` - `UnionNoOverwrite(winner Types, loser Types)`, where `winner` wins collisions.

Since your PR doesn't have such a method I could just add it here, you can factor yours out and I can rebase on top of it.

Reply:
Improved this